### ISSUE TITLE
Difference between MultistepAdamOptimizer and delayed_update

### ISSUE THREAD

**Issue by xueyouluo** [OpenNMT/OpenNMT-tf#470]
Hi, I am looking at the source code of accumulating gradients, I found:

In [MultistepAdamOptimizer](https://github.com/OpenNMT/OpenNMT-tf/blob/master/opennmt/optimizers/multistep_adam.py#L84) line 84, you have the gradients divided by n then do the update. But in the [delayed_update](https://github.com/OpenNMT/OpenNMT-tf/blob/master/opennmt/utils/optim.py#L202), I didn't found accumulated gradients divide by anything. I think the accumulated gradients should divide by the accum_count, right?

Could you tell me why there is a difference here?

**Comment by xueyouluo** [comment]
Emmm, if we didn't divide the accum_count, it equals to we have the learning rate multiply the accum_count. For small learning rate, I think it OK.

**Comment by guillaumekln** [comment]
Hi,

Good question! Averaging the gradients is indeed the correct way to do it.

The `delayed_update` approach is similar to OpenNMT-py where gradient accumulation is done be repeatedly calling forward which will simply accumulate (as is add) the gradients in the model. Would be interesting to compare.

**Comment by xueyouluo** [comment]
@guillaumekln Thank you for your reply. 

Another question is about the horovod. Since we need accumulate the gradient, you write DistributedOptimizer to wrapper the hvd.DistributedOptimizer. In this way, we can allreduce the accumulated gradients before apply_gradients. Horovod did this in compute_gradients. 

The question is, why did you set the average_gradients to False by default? If we didn't do the average, same thing happen like accum_count.

**Comment by guillaumekln** [comment]
I think this was to match the behavior of `delayed_update`. But if we find it's the same or better to average gradients (in terms of final model performance), we should do that everywhere.

**Comment by xueyouluo** [comment]
@guillaumekln Agree with you. What makes the model performance better is the right way. 😄
